### PR TITLE
Sdk improvements

### PR DESCRIPTION
**Pull request by z4kn4fein**
### Describe the purpose of your pull request

This PR applies the following improvements:
- During an ongoing fetch, all additional fetch requests will subscribe to the ongoing fetch's `CompletableFuture`.
- JSON deserialization happens only when the actual JSON differs from the cached one in memory.
- A new `logLevel()` configuration option was added, to let the user change the SDK's verbosity.
- The evaluation logs are staying on `info` level but they are now combined into a single log entry.
- A `getAllValues()` function was added that evaluates all flags & settings in a `Map<String, Object>`.
- New `flagOverrides()` builder option to support the reading of flags & settings from a local file or a `Map<String, Object>`.
#### Breaking Changes
- Fixed the names of the Polling modes: e.g: `PollingModes.AutoPoll()` -> `PollingModes.autoPoll()`.
- `maxWaitTimeForSyncCallsInSeconds` was replaced with `maxInitWaitTimeInSeconds` in case of auto poll to give a maximum amount of wait time for the first config JSON fetch. When it's reached without a result, the latest cached value will be returned.

### Requirement checklist (only if applicable)

- [x] I have covered the applied changes with automated tests.
- [x] I have executed the full automated test set against my changes.
- [x] I have validated my changes against all supported platform versions.
- [x] I have read and accepted the [contribution agreement](https://github.com/configcat/legal/blob/main/contribution-agreement.md).

### CODE REVIEW

File: src/main/java/com/configcat/ConfigCatClient.java [hunk 1]
```diff
@@ -127,15 +135,16 @@ public String getVariationId(String key, String defaultVariationId) {
...
+        if (key == null || key.isEmpty())
+            throw new IllegalArgumentException("'key' cannot be null or empty.");
 
         try {
-            return this.maxWaitTimeForSyncCallsInSeconds > 0
-                    ? this.getVariationIdAsync(key, user, defaultVariationId).get(this.maxWaitTimeForSyncCallsInSeconds, TimeUnit.SECONDS)
-                    : this.getVariationIdAsync(key, user, defaultVariationId).get();
+            return this.getVariationIdAsync(key, user, defaultVariationId).get();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```
Comment (kp-cat) [inline src/main/java/com/configcat/ConfigCatClient.java:144]:
```suggestion
            this.logger.error("Thread interrupted.", e);
            Thread.currentThread().interrupt();
```

File: src/main/java/com/configcat/ConfigCatClient.java [hunk 2]
```diff
@@ -70,30 +77,31 @@ public ConfigCatClient(String sdkKey) {
...
-            throw new IllegalArgumentException("Only String, Integer, Double or Boolean types are supported");
+            throw new IllegalArgumentException("Only String, Integer, Double or Boolean types are supported.");
 
         try {
-            return this.maxWaitTimeForSyncCallsInSeconds > 0
-                    ? this.getValueAsync(classOfT, key, user, defaultValue).get(this.maxWaitTimeForSyncCallsInSeconds, TimeUnit.SECONDS)
-                    : this.getValueAsync(classOfT, key, user, defaultValue).get();
+            return this.getValueAsync(classOfT, key, user, defaultValue).get();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```
Comment (kp-cat) [inline src/main/java/com/configcat/ConfigCatClient.java:101]:
```suggestion
            this.logger.error("Thread interrupted.", e);
            Thread.currentThread().interrupt();
```

**Comment by kp-cat** [comment]
Can you please add comments to the sample code where the info log level is set:
```
// Info level logging helps to inspect the feature flag evaluation process.
// Use the default Warning level to avoid too detailed logging in your application.
```

**Comment by z4kn4fein** [comment]
> Can you please add comments to the sample code where the info log level is set:
> 
> ```
> // Info level logging helps to inspect the feature flag evaluation process.
> // Use the default Warning level to avoid too detailed logging in your application.
> ```

Done

**Review by kp-cat — APPROVED** [review]

File: src/main/java/com/configcat/LocalFilePollingMode.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package com.configcat;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+class LocalFilePollingMode extends PollingMode {
```
Comment (kp-cat) [inline src/main/java/com/configcat/LocalFilePollingMode.java:8]: Should we implement `getPollingIdentifier()` functon here?

Comment (z4kn4fein) [reply src/main/java/com/configcat/LocalFilePollingMode.java:8]: `PollingMode` gives us a default implementation which is enough here I think as it's redundant with this mode.

File: src/main/java/com/configcat/ConfigCatClient.java [hunk 4]
```diff
@@ -44,20 +59,23 @@ private ConfigCatClient(String sdkKey, Builder builder) throws IllegalArgumentEx
...
+                        : builder.baseUrl,
                 hasCustomBaseUrl,
                 pollingMode.getPollingIdentifier());
 
         ConfigCache cache = builder.cache == null
                 ? new InMemoryConfigCache()
                 : builder.cache;
 
-        this.refreshPolicy = this.selectPolicy(pollingMode, cache, fetcher, this.logger, sdkKey);
+        this.refreshPolicy = this.overrideBehaviour == OverrideBehaviour.LOCAL_ONLY
```
Comment (kp-cat) [inline src/main/java/com/configcat/ConfigCatClient.java:76]: Maybe it's not needed to create the fetcher if the `overrideBehaviour` is `LOCAL_ONLY`

Comment (z4kn4fein) [reply src/main/java/com/configcat/ConfigCatClient.java]: done

**Review by kp-cat — APPROVED** [review]

**Pull request merged**